webapp/src/components/onboarding/LLMConfigurationForm.tsx

```diff
@@ -0,0 +1,564 @@
+	const isOpenAISelectedEmbeddingType = embeddingType.value === ModelType.OPENAI;
+
+	// const embeddingAPIKeyValue = embeddingAPIKey && embeddingAPIKey.startsWith('sk-') && embeddingAPIKey.length > 10
+	// 	? `${embeddingAPIKey.substring(0, 3)}${'X'.repeat(embeddingAPIKey.length - 3)}`
+	// 	: embeddingAPIKey;
+
+	// const llmAPIKeyValue = llmAPIKey && llmAPIKey.startsWith('sk-') && llmAPIKey.length > 10
+	// 	? `${llmAPIKey.substring(0, 3)}${'X'.repeat(llmAPIKey.length - 3)}`
+	// 	: llmAPIKey;
+
+	const LLMModelRequiredFields = Object.keys(ModelTypeRequirements[LLMType.value])
+		.filter(key => !ModelTypeRequirements[LLMType.value][key].optional)
+		.map(key => {
+			const label = key.replace(/_/g, ' ').replace(/\b\w/g, char => char.toUpperCase());
+			const placeholder = key.endsWith('key') ? 'Paste your API key' : `Enter the ${label}`;
+			return { name: key, label, placeholder };
+		});
+
+	const EmbeddingModelRequiredFields = Object.keys(ModelTypeRequirements[embeddingType?.value])
+		.filter(key => !ModelTypeRequirements[embeddingType?.value][key].optional)
+		.map(key => {
+			const label = key.replace(/_/g, ' ').replace(/\b\w/g, char => char.toUpperCase());
+			const placeholder = key.endsWith('key') ? 'Paste your API key' : `Enter the ${label}`;
+			return { name: `embedding_${key}`, label, placeholder };
+		});
+
+	const { isTablet, isMobile } = useResponsive();
+
+	const onSubmit = async (data: LLMConfigurationFormValues) => {
+		setSubmitting(true);
+
+		if (data.LLMModel.value) {
+			const body = {
+				_csrf: csrf,
+				resourceSlug,
+				name: data.LLMType.label,
+				model: data.LLMModel.value,
+				type: data.LLMType.value,
+				config: {
+					model: data.LLMModel.value,
+					...(data.api_key && { api_key: data.api_key }),
+					...(data.groq_api_key && { groq_api_key: data.groq_api_key }),
+					...(data.cohere_api_key && { cohere_api_key: data.cohere_api_key }),
+					...(data.base_url && { base_url: data.base_url })
+				}
+			};
+
+			const addedModel = await API.addModel(
+				body,
+				null,
+				res => {
+					toast.error(res);
+				},
+				null
+			);
+
+			const setDefaultModelBody = {
+				_csrf: csrf,
+				resourceSlug,
+				modelId: addedModel._id,
+				modelType: 'llm'
+			};
+
+			await API.setDefaultModel(
+				setDefaultModelBody,
+				null,
+				res => {
+					toast.error(res);
+				},
+				null
+			);
+		}
+
+		if (data.embeddingModel.value) {
+			const body = {
+				_csrf: csrf,
+				resourceSlug,
+				name: data.embeddingType.label,
+				model: data.embeddingModel.value,
+				type: data.embeddingType.value,
+				config: {
+					model: data.embeddingModel.value,
+					...(data.embedding_api_key && { api_key: data.embedding_api_key }),
+					...(data.embedding_cohre_api_key && { cohre_api_key: data.embedding_cohre_api_key }),
+					...(data.embedding_groq_api_key && { groq_api_key: data.embedding_groq_api_key }),
+					...(data.embedding_base_url && { base_url: data.embedding_base_url })
+				}
```

Comment:
See `ModelTypeRequirements.tsx` (used in the current model creation forms) for how it builds the "config" object dynamically.

How much effort would it take to change this form to do something similar? Or else this will become annoying to maintain -- we'll have to make changes to this form every time we add a model with new config parameters.

Reply:
you're right.. i've update the interface to make it more flexible but still keep some degree of control